MIRI LRS spectral extraction

karllark opened the pull request:
Works for the fixed and wavelength scaled boxcar extraction with and without background subtraction and PSF-weighed extraction.  For MIRI LRS.

karllark commented:
Issues:

1. How to integrate the jwst pipeline `extract_1d` code instead of my custom code for the boxcar extractions?  Note that custom code will still be needed for the PSF-based extractions.

2. How to normalize the PSF-based spectrum to the same level as the boxcar extracted spectrum?

3. Why is the jwst pipeline extracted spectrum (x1d) significantly lower than the extractions with my custom code?

karllark commented:
> 1. As mentioned, I hit an "HTTP Error 404: Not Found" error trying to retrieve the first data file. Please double check that you've followed the instructions [here](https://github.com/spacetelescope/style-guides/blob/master/guides/where-to-put-your-data.md#obtaining-static-box-links) for making the files downloadable from Box via direct link. I have minimal Box experience, so @robelgeda is probably your initial go-to here for help.

It looks like something has happened with the redirect *and* the link sharing.  The link sharing has been turned off on all my files.  In addition, the example for the redirect on the page given in the quote no longer works.

rosteen commented:
I was able to download the data from Box and run the rest of the notebook without any more problems. I don't think I have any more comments to add beyond the points I posted above, although I'm still not sure if I would prefer the functions defined in spextract.py to be defined in notebook cells instead, so that readers can more easily see them. @robelgeda , any thoughts on that?

robelgeda commented:
I would have to agree with @rosteen, we want to clearly show all the septs that went into the process. This is because some of the notebooks will be hosted on a notebook server that will not allow users to see the `*.py` files without doing a `cat` in the notebook itself. I think doing a subsection for each step should work fine. If this contents of the python files are needed for multiple notebooks, we will have to think of something claver.

karllark commented:
Email science review from Greg Sloan via email.

```
I have installed and tested it on both a MacBook and a Linux
system, and in both cases it runs smoothly.  The major impediment,
as I said before, is installation and configuring the environment.
Because I'm probably fairly representative of a Python neophyte in
the astronomy community, it's important that STScI educate people
accordingly.  My guess is that this is already happening!

The code runs fine.  The one major issue is that the extracted
spectra, in all their forms, are plotted in F_nu units, with the
implication that they are fully calibrated, but that is in fact not
the case, because the differ from each other in shape and in
overall flux level.

The extractions using the different methods should produce the same
general spectrum, but with differing levels of associated noise.

A more minor point is that the notebook should provide the user
with plotting options for the flux axis.  To better diagnose the
issues with the spectra, I'd prefer to plot them in Rayleigh-Jeans
units instead of F_nu units.  Perhaps a switch at the top of the
routine would facilitate this (plotting_units = 'Jy', 'RJ units',
etc.)

I noticed two minor hiccups, which I don't think are important but
should still be relayed.  In fact, it's the same hiccup twice.

In the two blocks of code that start with the comment
# extract the spectrum using the weight image

the following warning appears in a pink box:

.../anaconda3/envs/lrs_specex_test/lib/python3.8/site-packages/gwcs/wcs.py:131:
VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences
(which is a list-or-tuple of lists-or-tuples-or ndarrays with different
lengths or shapes) is deprecated. If you meant to do this, you must specify
'dtype=object' when creating the ndarray transforms =
np.array(self._pipeline[from_ind: to_ind])[:, 1].copy()

(Probably better to start each block with a different comment, too.)

The overall differences in the extracted spectra warrant a closer
inspection, but that should not diminish the fact that this software
looks good.  It's clean, and it runs smoothly.  I'm looking forward
to the chance to dig into it in more detail in the near future.
```

karllark commented:
I have updated the notebook to take into account the technical and science reviews.  I believe I have addressed all the comments except for creating a requirements.txt file with the specific versions needed for the notebook.  I could use instructions on how to gather this information.  And which packages should be included given a number are used for this notebook.

rosteen commented:
For the requirements file, I would include anything you explicitly import (matplotlib, numpy, astropy, jwst) plus jdaviz itself. You can get your installed versions with e.g. `pip list | grep matplotlib`. Packages that can be installed from pip should have lines in the requirements file like: `astropy == 4.0`. 

For packages that need to be retrieved from from github (jwst and jdaviz), the lines in the file should be e.g. `-e git+https://github.com/spacetelescope/jdaviz.git@0.1.dev537+g10bb72a.d20200626` where the url is the link to clone the repository and everything after the `@` is the result of your `pip list` call. We may not need to lock it down to a specific commit, in which case you can omit everything from the `@` onward in those lines - @robelgeda I forget, was the decision to lock jdaviz down to a specific commit for these notebooks or not?

rosteen commented on jdat_notebooks/MIRI_LRS_spectral_extraction/requirements:
```diff
@@ -0,0 +1,4 @@
+matplotlib==3.2.1
+numpy=1.18.4
+astropy=4.1
+jwst=0.16.1
```
```suggestion
numpy==1.18.4
astropy==4.0.1
git+https://github.com/spacetelescope/jwst@0.16.1
```

The requirements file runs successfully with these changes. Note that I changed `astropy` from 4.1 to 4.0.1, I assume that was the intent since there is no 4.1 release.

eteq commented:
Out-of-band it was communicated that this notebook is ready, but I just pushed up a commit that I *think* will get things working on the CI, so awaiting that to run.

eteq commented:
Huzzah it passed!  Merging, but @rosteen and @karllark may want to take a quick look at the modifications I made to the requirements file to get things to run in the CI just for possible future needs.
